ls1intum/Artemis
# `Notifications`: Fix unnecessary release notification when updating an exercise

### Checklist
#### General
#### Server

### Motivation, Context, and Description

This PR solves the found Bug : [Extra release notifications when editing released exercises #4238](https://github.com/ls1intum/Artemis/issues/4238)
The problem was that an "exercise released" notification was created and sent when the exercise was simply updated (no changes to the release date were made).
I added a check to make sure to only include such a notification if the release date changed during the exercise update.


I created a "decision matrix" to help me implement and cover every case so no more unexpected/uncovered behavior might slip through and create a bug. I also created adequate tests that check every possibility,
![Untitled](https://user-images.githubusercontent.com/65814168/141349614-117eddad-ae3b-475e-bf55-1650ad786487.png)

### Steps for Testing
Follow the steps written in the issue #4238
Play around to test if the legacy logic still works:
- create new exercises (of different types) with a release date : future/undefined/now/past
- reschedule an exercise release: e.g. Change the release date from (now: 10:00) (previouse release date: 10:05) to (new release date 10:02)
- change the exercise (when it was already released, e.g. change problem statement) and make sure no "exercise released" notification is created/send/shown

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage
I added a couple of new test cases that should cover all newly added logic to 100% in GroupNotificationService
Other parts of the system were not changed. I mostly added a new input method parameter.

Regarding the failing GitHub Actions:
I do not know why they fail.
I have not changed anything on the client side + the failling server test has nothing to do with my changes and works fine locally
![image](https://user-images.githubusercontent.com/65814168/141355180-6d4e8548-06ce-42ef-909f-ecd04280026a.png)
